cli: avoid logging during flag initialization

petermattis opened the pull request:
Avoid logging during flag initialization which happens before we've had
a chance to examine the --background flag. This was spuriously creating
a log file that was mostly empty.

Fixes #30120

Release note: None

petermattis committed:
cli: avoid logging during flag initialization

Avoid logging during flag initialization which happens before we've had
a chance to examine the --background flag. This was spuriously creating
a log file that was mostly empty.

Fixes #30120

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/30123)
<!-- Reviewable:end -->

petermattis commented:
This patches over the proximate cause of spurious log files. It would be nice to have a more robust way of not logging if `--background` is specified.

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> this is not a bad solution. 

I am a bit unsure about the root cause though. The log package should not be creating a log file until the `start` command has decided that a log file needs to be created. (i.e. until log.logDir is set) I am surprised the thing you found caused a log file to appear in the first place.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123#-:-LMJd29N9WHeWomf8Woi:b-nv1r3m)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

petermattis commented:
> I am a bit unsure about the root cause though. The log package should not be creating a log file until the start command has decided that a log file needs to be created. (i.e. until log.logDir is set) I am surprised the thing you found caused a log file to appear in the first place.

I'm happy to let you take this over and track down what is going on there.

knz commented:
I'd actually be glad to however it would need to wait a week or two. If you think that's inconvenient I'm happy to merge as-is

petermattis commented:
The log message is being generated when parsing the flag. Specifically, `bytesOrPercentageValue.Set` calls `memoryPercentResolver` which was logging a message. A log file was created, because cockroach was started with the `--log-dir` flag specified and this flag occurred on the command line before the flag which used `memoryPercentResolver`. Specifically, if you start cockroach with `--log-dir=foo --cache=25%` you encounter this problem. If you instead start cockroach with `--cache=25% --log-dir=foo` you do not. 

Seems like we want to have a layer of indirection for the `--log-dir` flag so that we parse it, but do not change the value in `util/log` until later in the startup process.

knz commented:
gotcha. Nice investigation, I would not have thought about argument ordering.

petermattis committed:
cli: only define the --log-{dir,file}* flags for the start/demo commands

Rework initialization for the `--log-{dir,file}*` flags so they are only
specified for the `start` and `demo` commands. Changed the handling of
the `--log-dir` flag so that configuration of the log dir used by
util/log is only performed after the final determination of the flag
value.

Release note (cli change): Only define --log-dir, --log-dir-max-size,
--log-file-max-size, and --log-file-verbosity for the start and demo
commands. Previously these flags were being defined on every command,
but rarely used or not functional.

petermattis commented:
Added a second commit to this PR which removes the `--log-{dir,file}*` flags for all commands except for `start` and `demo`. It also changes the handling of the `--log-dir` flag in a way that would have avoided the problem fixed by the first commit. PTAL.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cli/cliflags/flags.go, line 804 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30123#-LMOd2LmBbN0fowrVJEI:-LMOd2LmBbN0fowrVJEJ:b-nii11g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b115f168a869999e98f3262448b141a2db1ceb5/pkg/cli/cliflags/flags.go#L804)):*
> ```Go
> If non-empty, write log files in this directory. If empty, write log files to
> the <store-dir>/logs where <store-dir> is the directory of the first on disk
> store.
> ```

@piyush-singh or @jseldess the wording here and on the other help texts could be improved. I noticed that the default log directory stuff doesn't seem to be mentioned on our docs page: https://www.cockroachlabs.com/docs/v2.0/start-a-node.html#defaults


<!-- Sent from Reviewable.io -->

petermattis commented:
Ping. The second commit needs a review.

jseldess commented on pkg/cli/flags.go:
```diff
@@ -235,13 +235,17 @@ func init() {
+			logflags.LogFilesCombinedMaxSizeName,
+			logflags.LogFileVerbosityThresholdName:
+			// The --log-dir* and --log-file* flags are specified only for the
+			// `start` command.
```
```
`start` and `demo` commands.
```
?

petermattis replied:
Yep. Done.

jseldess reviewed (approved):
Sorry for the delay. LGTM, with a few nits.

jseldess commented on pkg/cli/cliflags/flags.go:
```diff
@@ -780,4 +780,34 @@ in the history of the cluster.`,
+		Name: "log-dir",
+		Description: `
+If non-empty, write log files in this directory. If empty, write log files to
+the <store-dir>/logs where <store-dir> is the directory of the first on disk
```
I'd remove `the` from `the <store-dir>/logs`.

petermattis replied:
Done.

petermattis reviewed (commented):
Thanks, Jesse. @knz Anything you'd like to add? Or @bdarnell who I believe had an opinion about the logging flags.

bdarnell reviewed (approved):
> Reviewed 1 of 8 files at r3, 7 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

knz reviewed (approved):
hey this change is actually pretty cool. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 1 of 8 files at r3, 6 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30123#-:-LMdV9px17XqpSYnpkU_:b9sdlrh)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

petermattis commented:
bors r=knz,bdarnell,jseldess

@jseldess I assume you've got the docs update for this on your plate (if there are any).

petermattis commented:
`acceptance/bank/node-restart` flake (which is on my plate to fix). Retrying.

bors r=knz,bdarnell,jseldess

jseldess commented:
@petermattis, yes, once this makes it into a release, our release notes script will pick up the change based on your release note text. And then we'll update the docs.
